extensions/interactions/NumberWithUnits/directives/NumberWithUnitsValidationService.js

```diff
@@ -0,0 +1,135 @@
+// Copyright 2018 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Validator service for the number with units interaction.
+ */
+
+oppia.factory('NumberWithUnitsValidationService', [
+  'WARNING_TYPES', 'baseInteractionValidationService',
+  'NumberWithUnitsObjectFactory',
+  function(WARNING_TYPES, baseInteractionValidationService,
+      NumberWithUnitsObjectFactory) {
+    return {
+      getCustomizationArgsWarnings: function(customizationArgs) {
+        return [];
+      },
+      getAllWarnings: function(
+          stateName, customizationArgs, answerGroups, defaultOutcome) {
+        var warningsList = [];
+
+        warningsList = warningsList.concat(
+          this.getCustomizationArgsWarnings(customizationArgs));
+
+        var checkEquality = function(earlierRule, laterRule) {
+          earlierInput = NumberWithUnitsObjectFactory.fromDict(
+            earlierRule.inputs.f);
+          laterInput = NumberWithUnitsObjectFactory.fromDict(
+            laterRule.inputs.f);
+
+          if (earlierInput.type === laterInput.type) {
+            if (earlierInput.type === 'fraction') {
+              if (earlierInput.fraction.tofloat() !== laterInput.fraction.
+                toFloat()) {
+                return false;
+              }
+
+              if (!angular.equals(earlierInput.units.toDict(), laterInput.units.
+                toDict())) {
+                return false;
+              }
+              return true;
+            } else if (earlierInput.type === 'real') {
+              return earlierInput.real === laterInput.real && angular.equals(
+                earlierInput.units.toDict(), laterInput.units.toDict());
+            }
+          }
+          return false;
+        };
+
+        var checkEquivalency = function(earlierRule, laterRule) {
```

Comment:
This code seems pretty similar to the `isEquivalentTo` rule in the rulesService. Better to use the rules service here or make a separate factory which will be used by rules service and this service.

Reply:
Correct. But can I use rulesService in validation service? Else, we can stick  to this approach.

Reply:
I don't see a problem in doing that.